include/matx/transforms/cub.h

```diff
@@ -1092,6 +1096,163 @@ inline void ExecSort(OutputTensor &a_out,
   size_t temp_storage_bytes = 0;
 };
 
+struct CustomArgMaxCmp
+{
+  template <typename T>
+  __MATX_DEVICE__ __MATX_HOST__ __MATX_INLINE__ T operator()(const T &a, const T &b) const {
+    return thrust::get<1>(a) < thrust::get<1>(b) ? b : a;
+  }
+};
+
+template <typename OutputTensor, typename TensorIndexType, typename InputOperator, CUBOperation_t op, typename CParams = EmptyParams_t>
+class matxCubTwoOutputPlan_t {
+  using T1 = typename InputOperator::value_type;
+
+public:
+  matxCubTwoOutputPlan_t(OutputTensor &a_out, TensorIndexType &aidx_out, const InputOperator &a, const CParams &cparams, const cudaStream_t stream = 0) :
+    cparams_(cparams)
+  {
+#ifdef __CUDACC__
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL)
+
+    if constexpr (op == CUB_OP_REDUCE_ARGMAX) {
+      ExecArgMax(a_out, aidx_out, a, stream);
+    }
+    else {
+      MATX_THROW(matxNotSupported, "Invalid CUB operation");
+    }
+
+    // Allocate any workspace needed by underly CUB algorithm
+    matxAlloc((void **)&d_temp, temp_storage_bytes, MATX_ASYNC_DEVICE_MEMORY,
+              stream);
+#endif
+  }
+  
+  static auto GetCubParams([[maybe_unused]] OutputTensor &a_out,
+                           [[maybe_unused]] TensorIndexType &aidx_out,
+                                  const InputOperator &a,
+                                  cudaStream_t stream)
+  {
+    CubParams_t params;
+
+    for (int r = 0; r < InputOperator::Rank(); r++) {
+      params.size.push_back(a.Size(r));
+    }
+
+    params.op = op;
+    if constexpr (OutputTensor::Rank() > 0)
+    {
+      params.batches = TotalSize(a_out);
+    }
+    else
+    {
+      params.batches = 1;
+    }
+    params.dtype = TypeToInt<T1>();
+    params.stream = stream;
+
+    return params;
+  }
+
+  /**
+   * Execute an argmax on a tensor
+   *
+   * @note Views being passed must be in row-major order
+   *
+   * @tparam OutputTensor
+   *   Type of output tensor
+   * @tparam TensorIndexType
+   *   Type of the output index tensor
+   * @tparam InputOperator
+   *   Type of input tensor
+   * @param a_out
+   *   Output tensor
+   * @param aidx_out
+   *   Output index tensor
+   * @param a
+   *   Input tensor
+   * @param stream
+   *   CUDA stream
+   *
+   */
+  inline void ExecArgMax(OutputTensor &a_out,
+                         TensorIndexType &aidx_out,
+                         const InputOperator &a,
+                         const cudaStream_t stream)
+  {
+#ifdef __CUDACC__
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL)
+
+    CustomArgMaxCmp cmp_op;
+    auto initial_value = cuda::std::tuple<matx::index_t,typename InputOperator::value_type>(-1, std::numeric_limits<typename InputOperator::value_type>::lowest());
+
+    if constexpr (OutputTensor::Rank() > 0) {
+      auto a_iter = matx::RandomOperatorOutputIterator{a};
```

Comment:
Why does the input use an output iterator?

Reply:
This is a thrust `reference operator*() const` need.  I've modified the code to add a RandomOperatorThrustIterator class that meets Thrust's unique needs (which are slightly different from CUB).